File: kalman_1d.py
import numpy
import pylab
import scipy

# intial parameters
n_iter = 50
sz = (n_iter,) # size of array
y0 = 100  # initial value
g = 9.81
dt = 0.2
ytruth = numpy.zeros(sz)
for i in range(0,n_iter):
    ytruth[i] = y0 - g*i*dt - g*i*dt**2/2.
ymeas = numpy.random.normal(ytruth,2) # observations

Q = 0 #1e-5 # process variance

# allocate space for arrays
yhat=numpy.zeros(sz)      # a posteri estimate of x
yhatminus=numpy.zeros(sz) # a priori estimate of x

R = 1**2 # estimate of measurement variance, change to see effect

# intial guesses
yhat[0] = 100.
P = 1.

for k in range(1,n_iter):
    # time update
     yhatminus[k] =  yhat[k-1] - g * dt - g*dt**2/2.
     Pminus = P #+Q

     # measurement update
     K = Pminus/( Pminus+R )
     yhat[k] = yhatminus[k]+K*(ymeas[k]-yhatminus[k])
     P = (1-K)*Pminus

pylab.figure()
pylab.plot(ytruth,'r-',color='g',label='truth value')
pylab.plot(ymeas,'k+',label='noisy measurements')
pylab.plot(yhat,'b-',label='a posteri estimate')
pylab.plot(yhatminus,'y-',label='a priori estimate')
pylab.legend()
pylab.xlabel('Iteration')
pylab.ylabel('Height')
pylab.show()

File: kalman1d.m
% -*- comment-start: "%"; -*-

clear;

global g dt;

N =  20;
dt = 0.2;
g = 9.81;

%%%% FIRST START WITH CREATING TRUE TRAJECTORY AND MEASURED (SMEARED) ONE
y0true = 100;
ytrue = zeros(1,N);
ymeas = zeros(1,N);
sigma = 2;
for i=1:N
    ytrue(i) = y0true - (i-1)*g*dt - (i-1)*g*dt^2/2;
    ymeas(i) = normrnd(ytrue(i),sigma);
end

%%%% CREATE INITIAL
%state vector
x = zeros(1,N);
x(1) = ymeas(1)*0.5 ;

% covariance
P = 10;

%measurement variance
R = 1;

%%%% KALMAN
x_= zeros(1,N);
for t=2:N
    x_(t) =x(t-1) - g*dt - g*dt^2/2;
    P_ = P;
    K = P_/(P_+R);
    x(t) = x_(t) + K*(ymeas(t)-x_(t));
    P =(1-K)*P_;
end

figure;
plot(ytrue, 'b',ymeas, 'rx', x_, 'g+', x, 'k+');
xlabel('Iteration');
ylabel('Position');
legend('True trajectory', 'Measured position', 'Prediction','Kalman');
print -dpng kalman1d.png

